### PR TITLE
Remove sorts when replacing sort aggregate if possible

### LINKED ISSUE #601: [FEA] Optimize unnecessary sorts when replacing SortAggregate
**Is your feature request related to a problem? Please describe.**
I noticed a query that had a GPU plan with this sequence of nodes somewhere in it:
  GpuSort --> GpuHashAggregate --> GpuColumnarExchange --> GpuSort --> GpuHashAggregate

GpuHashAggregate makes no assumptions on the order of its input, so the two sorts aren't accomplishing anything useful.  It looks like the original CPU plan used a sort-based aggregate and the plugin blindly replaced the sort with GpuSort and SortAggregate with HashAggregate.  If the plugin is going to use HashAggregate to replace SortAggregate then it should remove the unnecessary sorts.

**Describe the solution you'd like**
The plugin removes unnecessary sorts in the plan when replacing a sort aggregate with a hash aggregate.  This is similar to the behavior the plugin already performs when replacing a SortMergeJoin with a GpuShuffledHashJoin.

Note that a sort may be required _after_ the aggregate if subsequent nodes in the plan need the output ordering guaranteed by SortAggregate.  We should be able to leverage the same "insert sorts when needed" logic that is being performed when replacing SortMergeJoin with GpuShuffledHashJoin.

**Describe alternatives you've considered**
We could try to leverage libcudf's sort-based aggregate functionality, but we probably only want to leverage it when we know the data happens to already be sorted on the grouping keys before the aggregate.  I don't think we want to explicitly sort on the keys, as libcudf will already do this if necessary when performing an aggregate that must be performed using sorted data.

### PR DESCRIPTION
**Pull request by revans2**
This fixes #601 

The code will remove the sort iff the sort could run on the GPU just like with the sort merge join, and there are no first or last aggregates, which become deterministic( or at least a lot more deterministic) when sort is involved, but are not otherwise.

### CODE REVIEW

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala [hunk 1]
```diff
@@ -246,6 +246,34 @@ class GpuSortAggregateMeta(
             s"'complete', or 'all'")
       }
     }
+
+    // make sure this is last check - if this is SortAggregate, the children can be Sorts and we
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala:250]:
```suggestion
    // make sure this is the last check - if this is SortAggregate, the children can be Sorts and we
```

**Comment by revans2** [comment]
I manually verified that the code is removing the sort.  If someone really wants me to add in a unit test for this I can, but I don't think it is necessary.

Just FYI for the simple aggregate I ran removing the sort saved about 25% of the total run time.

**Comment by revans2** [comment]
build

**Review by jlowe — APPROVED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala [hunk 2]
```diff
@@ -246,6 +246,34 @@ class GpuSortAggregateMeta(
             s"'complete', or 'all'")
       }
     }
+
+    // make sure this is the last check - if this is SortAggregate, the children can be Sorts and we
```
Comment (kuhushukla) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala:250]: Super nitty nit : Make *

**Review by kuhushukla — APPROVED** [review]
Super minor nit which may be skipped. And maybe we can fix the typo in the PR title as well but approving as the change looks good.

**Comment by tgravescs** [comment]
I will be interested to test the adobe query after this goes in to see if it makes a difference for that

**Pull request merged**